computestorage/helpers.go

```diff
@@ -0,0 +1,173 @@
+	var (
+		baseVhdPath = filepath.Join(layerPath, vhdName+"-base.vhdx")
+		diffVhdPath = filepath.Join(layerPath, vhdName+".vhdx")
+		hivesPath   = filepath.Join(layerPath, "Hives")
+		layoutPath  = filepath.Join(layerPath, "Layout")
+	)
+
+	// We need to remove the hives directory and layout file as `SetupBaseOSLayer` fails if these files
+	// already exist. `SetupBaseOSLayer` will create these files internally. We also remove the base and
+	// differencing disks if they exist in case we're asking for a different size.
+	if _, err := os.Stat(hivesPath); err == nil {
+		os.RemoveAll(hivesPath)
+	}
+	if _, err := os.Stat(layoutPath); err == nil {
+		os.RemoveAll(layoutPath)
+	}
+	if _, err := os.Stat(baseVhdPath); err == nil {
+		os.RemoveAll(baseVhdPath)
+	}
+	if _, err := os.Stat(diffVhdPath); err == nil {
+		os.RemoveAll(diffVhdPath)
+	}
+
+	createParams := &vhd.CreateVirtualDiskParameters{
+		Version: 2,
+		Version2: vhd.CreateVersion2{
+			MaximumSize:      sizeInGB * 1024 * 1024 * 1024,
+			BlockSizeInBytes: defaultVHDXBlockSizeInMB * 1024 * 1024,
+		},
+	}
+	handle, err := vhd.CreateVirtualDisk(baseVhdPath, vhd.VirtualDiskAccessNone, vhd.CreateVirtualDiskFlagNone, createParams)
+	if err != nil {
+		return errors.Wrap(err, "failed to create vhdx")
+	}
+
+	defer func() {
+		if err != nil {
+			syscall.CloseHandle(handle)
+			os.RemoveAll(baseVhdPath)
+			if os.Stat(diffVhdPath); err == nil {
+				os.RemoveAll(diffVhdPath)
+			}
+		}
+	}()
+
+	if err = FormatWritableLayerVhd(ctx, windows.Handle(handle)); err != nil {
+		return err
+	}
+	// Base vhd handle must be closed before calling SetupBaseLayer in case of Container layer
+	if err = syscall.CloseHandle(handle); err != nil {
+		return errors.Wrap(err, "failed to close vhdx handle")
+	}
+
+	options := OsLayerOptions{
+		Type: OsLayerTypeContainer,
+	}
+
+	// SetupBaseOSLayer expects an empty vhd handle for a container layer and will
+	// error out otherwise.
+	if err = SetupBaseOSLayer(ctx, layerPath, 0, options); err != nil {
+		return err
+	}
+	// Create the differencing disk that will be what's copied for the final rw layer
+	// for a container.
+	if err = vhd.CreateDiffVhd(diffVhdPath, baseVhdPath, defaultVHDXBlockSizeInMB); err != nil {
+		return errors.Wrap(err, "failed to create differencing disk")
+	}
+
+	if err = security.GrantVmGroupAccess(baseVhdPath); err != nil {
+		return errors.Wrapf(err, "failed to grant vm group access to %s", baseVhdPath)
+	}
+	if err = security.GrantVmGroupAccess(diffVhdPath); err != nil {
+		return errors.Wrapf(err, "failed to grant vm group access to %s", diffVhdPath)
+	}
+	return nil
+}
+
+// SetupUtilityVMBaseLayer is a helper to setup a UVMs scratch space. It will create and format
+// the vhdx inside and the size is configurable by the sizeInGB parameter.
+func SetupUtilityVMBaseLayer(ctx context.Context, uvmPath string, vhdName string, sizeInGB uint64) (err error) {
```

Comment:
We need to change the signature of this function to take one more argument named `vhdCreationPath` or something like that. In case of cimfs the uvmPath (which contains all the layer files) is going to be a path to the mounted read only cim. So we can not create any VHDs on that path.  Or maybe there is some other better way in which we can refactor these two function to include support for this.

Reply:
Done!